File: Assignment4/2.c
// Write a program to convert a string into a linked list of characters.
#include <stdio.h>
#include<stdlib.h>

struct Node{
    int data;
    struct Node* next;
};

struct Node* createNode(int data){
    struct Node* newNode = (struct Node*)malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

void printList(struct Node* head){
    struct Node* temp = head;
    while(temp != NULL){
        printf("%c-> ",temp->data);
        temp = temp->next;
    }
    printf("NULL");
}

int main() {
    char str[100];
    printf("Enter a string: ");
    scanf("%s",str);

    struct Node* head = NULL;
    struct Node* tail = NULL;

    for(int i=0; str[i] != '\0'; i++){
        struct Node* newNode = createNode(str[i]);
        if(head == NULL){
            head = newNode;
            tail = newNode;
        }
        else{
            tail->next = newNode;
            tail = newNode;
        }
    }

    printf("Linked list of characters:\n");
    printList(head);

    return 0;
}

File: Week-4/2.py
# Write a program to convert a string into a linked list of characters.
class Node:
    def __init__(self,data):
        self.data = data
        self.next = None

class LinkedList:
    def __init__(self):
        self.head = None

    def insert_end(self,data):
        new_node = Node(data)
        if not self.head:
            self.head = new_node
            return
        temp = self.head
        while temp.next:
            temp = temp.next
        temp.next = new_node

    def print_list(self):
        temp = self.head
        while temp:
            print(temp.data,end=" -> ")
            temp = temp.next
        print("NULL")

s = input("Enter a string: ")
ll = LinkedList()

for ch in s:
    ll.insert_end(ch)
ll.print_list()